### ISSUE TITLE
tmpnam() causes bus error crash (NULL pointer access)

### ISSUE THREAD

**Issue by czietz** [freemint/mintlib#37]
While porting some software, I found a bug in [tmpnam()](https://github.com/freemint/mintlib/blob/f5c219b1d868dd3f3249c4cf89a50f874f08fd2e/stdio/tmpnam.c#L22) or rather in [__get_tmpdir()](https://github.com/freemint/mintlib/blob/f5c219b1d868dd3f3249c4cf89a50f874f08fd2e/misc/diraccess.c#L40).

It can be reproduced by a minimal example, e.g.
```
/* tmpnam example */
#include <stdio.h>

int main ()
{
  char buffer [L_tmpnam];
  char * pointer;

  tmpnam (buffer);
  printf ("Tempname #1: %s\n",buffer);

  pointer = tmpnam (NULL);
  printf ("Tempname #2: %s\n",pointer);

  return 0;  
}
```
(source: http://www.cplusplus.com/reference/cstdio/tmpnam/)

Note that tmpnam() calls __get_tmpdir (NULL, 0). Therefore, executions ends up here: https://github.com/freemint/mintlib/blob/f5c219b1d868dd3f3249c4cf89a50f874f08fd2e/misc/diraccess.c#L58
... doing a strcmp() with NULL. (trydir is initialized to NULL.) Since there is no code to handle this special case in strcmp(), this crashes with a bus error in user mode.

I'm not really sure what the strcmp() is meant for, which is why I cannot propose a patch offhandedly.

**Comment by th-otto** [comment]
On Mittwoch, 17. April 2019 20:52:18 CEST Christian Zietz wrote:
> While porting some software, I found a bug in tmpnam() or rather in
> __get_tmpdir().

Wow. Looks like nobody ever used that function, since it will always crash. 
Thanks for spotting that, should hopefully be fixed with commit b3f7f23 

> I'm not really sure what the strcmp() is meant for

I think it is just to avoid a 2nd stat() on a name that has already been 
checked.

BTW.: that function was rather bogus by all means. After calls to 
__diraccess() the function returned  on failure instead of success(), same for 
__quickstat(). I doubt that this has ever been tested.

**Comment by czietz** [comment]
I can confirm that it does not crash anymore. However, it does not work, either, at least under TOS. tmpnam() returns NULL because __get_tmpdir() returns NULL, even though it should accept "\TMP" as temporary directory even under TOS: https://github.com/freemint/mintlib/blob/b3f7f23d1db3b235d68079b9ae33b92c21227e1d/misc/diraccess.c#L61-L62

I think, though, that this is another problem in mintlib and will open a separate issue.
